File: ts/example/src/tests/complex_tests.ts
import assert from "assert";
import {ActualBothCommandAndQueryUser, ActualCallbackUser, ActualManyMethodsUser} from "../complex_user";
import {
  BothCommandAndQuerySpy,
  anyBothCommandAndQueryStub,
  ExampleCallbackStub,
  ExampleManyMethodsSpy,
  ExampleManyMethodsStub,
} from "./complex_doubles";



export function testUsingBothCommandAndQuerySpy() {
  const spy = new BothCommandAndQuerySpy(anyBothCommandAndQueryStub());

  const sut = new ActualBothCommandAndQueryUser(spy);

  sut.doSomething();

  assert.strictEqual(spy.calledArgs.length, 1);
}


export function testUsingManyMethodsStub() {
  const stub = new ExampleManyMethodsStub(123, 456);

  const sut = new ActualManyMethodsUser(stub);

  const actual = sut.doSomething();
  assert.strictEqual(actual, 456);
}


export function testUsingManyMethodsSpy() {
  const stub = new ExampleManyMethodsSpy(
    new ExampleManyMethodsStub(123, 456)
  );

  const sut = new ActualManyMethodsUser(stub);

  const actual = sut.doSomething();
  assert.strictEqual(actual, 456);
}


export function testUsingCallbackStub() {
  const stub = new ExampleCallbackStub();

  const sut = new ActualCallbackUser(stub);

  sut.doSomething();

  stub.fireCallback("Hello, World!");

  assert.strictEqual(sut.isCalled, true);
}

File: swift/example/Tests/VanillaMockExampleTests/ComplexTests.swift
import XCTest
@testable import VanillaMockExample



class ComplexTests: XCTestCase {
    func testUsingBothCommandAndQuerySpy() {
        let spy = BothCommandAndQuerySpy()

        let sut = ActualBothCommandAndQueryUser(cmdAndQuery: spy)

        _ = sut.doSomething()

        XCTAssertEqual(spy.callArgs.count, 1)
    }


    func testUsingManyMethodsStub() {
        let stub = ExampleManyMethodsStub(barResult: 123, bazResult: 456)

        let sut = ActualManyMethodsUser(manyMethods: stub)

        let actual = sut.doSomething()

        XCTAssertEqual(actual, 456)
    }


    func testUsingManyMethodsSpy() {
        let spy = ExampleManyMethodsSpy(dynamicallyInheriting: ExampleManyMethodsStub(barResult: 456, bazResult: 789))

        let sut = ActualManyMethodsUser(manyMethods: spy)

        _ = sut.doSomething()

        XCTAssertEqual(spy.callArgs, [.foo(123), .bar, .baz(456)])
    }


    func testUsingCallbackStub() {
        let stub = ExampleCallbackStub()

        let sut = ActualCallbackUser(stub)

        sut.doSomething()
        stub.fireCallback(x: "Hello, World!")

        XCTAssertEqual(sut.isCalled, true)
    }
}